libr/include/r_util/r_buf.h

```diff
@@ -8,84 +8,104 @@ extern "C" {
 
 // TODO: choose whether the _at operations should preserve the current seek or not
 
-#define R_BUF_CUR UT64_MAX
+#define R_BUF_SET 0
+#define R_BUF_CUR 1
+#define R_BUF_END 2
+
+typedef struct r_buf_t RBuffer;
+
+typedef bool (*RBufferInit)(RBuffer *b, const void *user);
+typedef bool (*RBufferFini)(RBuffer *b);
+typedef int (*RBufferRead)(RBuffer *b, ut8 *buf, size_t len);
+typedef int (*RBufferWrite)(RBuffer *b, const ut8 *buf, size_t len);
+typedef ut64 (*RBufferGetSize)(RBuffer *b);
+typedef bool (*RBufferResize)(RBuffer *b, ut64 newsize);
+typedef int (*RBufferSeek)(RBuffer *b, st64 addr, int whence);
+typedef ut8 *(*RBufferGetWholeBuf)(RBuffer *b, ut64 *sz);
+typedef void (*RBufferFreeWholeBuf)(RBuffer *b);
+typedef RList *(*RBufferNonEmptyList)(RBuffer *b);
+
+typedef struct r_buffer_methods_t {
+	RBufferInit init;
+	RBufferFini fini;
+	RBufferRead read;
+	RBufferWrite write;
+	RBufferGetSize get_size;
+	RBufferResize resize;
+	RBufferSeek seek;
+	RBufferGetWholeBuf get_whole_buf;
+	RBufferFreeWholeBuf free_whole_buf;
+	RBufferNonEmptyList nonempty_list;
+} RBufferMethods;
 
 typedef struct r_buf_t {
-	ut8 *buf_priv;
-	ut64 length_priv;
-	st64 cur_priv;
-	// FIXME: some direct accesses to base_priv still exist unfortunately
-	ut64 base_priv;
-	RMmap *mmap_priv;
-	bool empty_priv;
-	bool ro_priv; // read-only
-	int fd_priv;
+	const RBufferMethods *methods;
```

Comment:
so much c++

Reply:
eheh true. the alternative was to have a switch case in every function, which seemed more messy.